### ISSUE TITLE
Predictable overscanning

### ISSUE THREAD

**Issue by fabiospampinato** [petyosi/react-virtuoso#188]
The way overscanning is currently implemented is counterintuitive for me, for a few reasons:

1. The viewport seems to be taken into account too, instead I'd expect overscanning to just add the provided overscan value to the visible height of the list basically. Having the viewport dimension hardcoded in the logic makes it difficult to fine-tune overscanning for our needs. Basically I'd expect Virtuoso to work at the list-level, not at the app-level, whatever the height/width of the window is it shouldn't matter at all.
2. The overscan value is doubled automatically, which I wouldn't have expected and I'm not sure what the reason behind that is.
3. Overscanning seems to work in chunks, which probably explains my previous point, but I'd expect Virtuoso to render the minimum number of items possible given the provided configuration, overscanning in chunks goes against that, so maybe it isn't worth it after all?
4. Since it's not currently possible to explicitly provide an overscan value for both scrolling directions I expected the current single value to work for both directions, but that's not the case, only the current scrolling direction takes advantage of overscanning it seems. I'd like the ability to configure an overscan value for both directions (last scrolling direction and opposite).
5. Since the used overscan depth is different depending on the direction a list is being navigated I'd expect an API to hint Virtuoso about changes in direction. Virtuoso can currently detect changes in direction when the list is scrolled, but it's possible to change direction without scrolling the list too, for example by highlighting the previous/next item in the list the user could be navigating up/down without scrolling the list.
6. Because of points 4 and 5 it's possible that only a single item outside the visible portion of the list has already been added to the DOM in one of the two scrolling directions, which kind of goes against having overscanning support in the first place IMHO.

Basically overscanning currently feels somewhat unpredictable and not fine-tunable enough.

**Comment by petyosi** [comment]
Thanks for your time to write that down :). Let's tackle the easy parts first. 

Point 1. in the codebase, the "viewport" is the element that encompasses the list. The component does not query the window size, so that's working as expected. 

Point 2. The `* 2` thing was a stupid idea. It is removed in the version I am working on, which is not yet released. 

Point 3. This is probably what we need to double down on. The overscan property is meant to do exactly that - "chunk" the renders, rendering less frequently (but more expensively).  It's not a performance silver bullet by any means. I see no point in overscan just acting as an artificial increase of the viewport size - but I might be wrong in my understanding.  

Point 4. Agreed. In the next version, the prop accepts `{ top: number, bottom: number }` configuration. 

Point 5. That's a cool idea, but I need to think about it further.

Point 6. Agreed. The change from Point 4 should address that.

Thank you once again for taking your time with this!

**Comment by fabiospampinato** [comment]
> Point 1. in the codebase, the "viewport" is the element that encompasses the list. The component does not query the window size, so that's working as expected.

My bad 👍 

> I see no point in overscan just acting as an artificial increase of the viewport size - but I might be wrong in my understanding.

I think that's how react-window does it. I'm no expert but I think just rendering items outside the viewport is important for ensuring tabbing works for example, other than that I think we just want to minimize glitches while scrolling and maximize performance, whichever approach does best at that I'm in favor of, if I had to guess I would guess that chunking is more expensive but I may very well be wrong.

> Point 4. Agreed. In the next version, the prop accepts { top: number, bottom: number } configuration.

Nice! 

When deciding which values to assign to those properties though my thinking is that either they should have the same value, in which case just providing Virtuoso a single number like it's currently possible would be a cleaner approach, or the direction that's currently being scrolled into should have a higher overscan value, in which case I would prefer to have properties like "main" and "opposite" rather than "top" and "bottom".

Like ideally I think I would set something like this: `overscan={{ main: 500, opposite: 200 }}`, and then Virtuoso would assign the "main" value to the current scroll direction, the "opposite" value to the opposite scroll direction, and maybe assign the "main" value to both directions if it isn't quite sure which way the list is likely to get scrolled next. Does that sound reasonable? 🤔 

I'm looking forward to trying the next version, it sounds like it's going to be a major improvement 💪

**Comment by petyosi** [comment]
Nice suggestion on main / opposite! I will think over it.

**Comment by petyosi** [comment]
@fabiospampinato v1 beta is available, and it includes the main/opposite improvement you suggested - thank you very much ;). Please test it and let me know how it works for you! #203 has the instructions and refers to the breaking changes.
